pyproject.toml

```diff
@@ -67,13 +52,137 @@ dev = [
 
 [build-system]
 build-backend = "mesonpy"
-# Also keep in sync with optional dependencies above.
+# Also keep in sync with dependency groups below.
 requires = [
     "meson-python>=0.13.1,<0.17.0",
     "pybind11>=2.13.2,!=2.13.3",
     "setuptools_scm>=7",
 ]
 
+[dependency-groups]
+build = [
+    # Should be the same as `[project] dependencies` above.
+    "contourpy >= 1.0.1",
+    "cycler >= 0.10",
+    "fonttools >= 4.22.0",
+    "kiwisolver >= 1.3.1",
+    "numpy >= 1.25",
+    "packaging >= 20.0",
+    "pillow >= 9",
+    "pyparsing >= 3",
+    "python-dateutil >= 2.7",
+
+    # Should be the same as `[build-system] requires` above.
+    "meson-python>=0.13.1,<0.17.0",
+    "pybind11>=2.13.2,!=2.13.3",
+    "setuptools_scm>=7",
+    # Not required by us but setuptools_scm without a version, so _if_
+    # installed, then setuptools_scm 8 requires at least this version.
+    # Unfortunately, we can't do a sort of minimum-if-installed dependency, so
+    # we need to keep this for now until setuptools_scm _fully_ drops
+    # setuptools.
+    "setuptools>=64",
+]
+dev = [
+    {include-group = "build"},
+    {include-group = "doc"},
+    {include-group = "test"},
+    {include-group = "test-extra"},
+    "ruff",
```

Comment:
`pre-commit` should be added here, and perhaps `ruff` dropped since running pre-commit runs ruff.